### PR TITLE
[https://nvbugs/5302040][feat] Add whisper support (Bert Attention on SM100 and GPTAttention for cross attention on SM100)

### PR DESCRIPTION
**Pull request by wu6u3tw**
# [nvbugs/5302040] feat. Add whisper support (Bert Attention on SM100 and GPTAttention for cross attention on SM100)

## Description
- Add bert attention plugin for Blackwell: 
    implement with fmhaDispatcher
  
- Add same seq_len causal mask support for gpt attention on cross attention:
    fix mask type and fix mask plugin layer

Please explain the issue and the solution in short.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved GPU compatibility for BERT/GPT attention across architectures.
  * Limited use of custom/packed attention masks to older GPUs (SM < 100), preventing issues on newer devices.
  * Removed an unnecessary hardware guard, enabling the attention plugin on additional GPUs.

* **Refactor**
  * Internal attention kernel dispatch and initialization updated for more robust execution and support checks.
  * No public API changes; behavior remains the same for supported configurations.

### CODE REVIEW

**Comment by nvzhihanj** [comment]
@PerkzZheng @kaiyux can you review or suggest the right PIC to review this PR?
This is to enable Whisper submission for MLPerf 5.1 B200 - adding support for SM100

**Comment by nvzhihanj** [comment]
@wu6u3tw please run precommit hook locally and call bot run in the comment section

**Review by PerkzZheng — COMMENTED** [review]
Other changes LGTM. BTW, are there any tests added yet to guard this (Whisper) ?

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp [hunk 1]
```diff
@@ -520,7 +520,14 @@ int BertAttentionPlugin::enqueueImpl(nvinfer1::PluginTensorDesc const* inputDesc
                 cudaMemsetAsync(fmhaParams.outputPtr, 0, ring_block_output_size, stream);
                 cudaMemcpyAsync(fmhaParams.tileCounterPtr, fmha_scheduler_counter_h, sizeof(uint32_t),
                     cudaMemcpyHostToDevice, stream);
-                mFMHARunner->run(fmhaParams);
+                if (tensorrt_llm::common::getSMVersion() >= 100 && tensorrt_llm::common::getSMVersion() < 120)
```
[the remark this answers is not in the document]
Comment (wu6u3tw) [reply cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:523]: Let me try and test it locally.

File: tensorrt_llm/functional.py [hunk 2]
```diff
@@ -5804,7 +5805,7 @@ def gpt_attention(
     if attention_mask is not None and mask_type == AttentionMaskType.custom_mask:
         # useFullCustomMask
         plug_inputs += [attention_mask]
-    if attention_packed_mask is not None:
+    if attention_packed_mask is not None and get_sm_version() < 100:
```
Comment (nvzhihanj) [inline tensorrt_llm/functional.py:5808]: Same here, I am guessing these 2 are causing the CI issues.

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.h [hunk 3]
```diff
@@ -115,6 +116,7 @@ class BertAttentionPlugin : public BasePlugin
 
     // The default copy constructor will leave them as nullptr. clone() shall initialize it.
     UniqPtrWNullCopy<tensorrt_llm::kernels::FusedMHARunnerV2> mFMHARunner;
+    UniqPtrWNullCopy<tensorrt_llm::kernels::FmhaDispatcher> mFmhaDispatcher;
```
[the remark this answers is not in the document]
Comment (wu6u3tw) [reply cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.h:117]: cleaned

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp [hunk 4]
```diff
@@ -520,7 +520,7 @@ int BertAttentionPlugin::enqueueImpl(nvinfer1::PluginTensorDesc const* inputDesc
                 cudaMemsetAsync(fmhaParams.outputPtr, 0, ring_block_output_size, stream);
                 cudaMemcpyAsync(fmhaParams.tileCounterPtr, fmha_scheduler_counter_h, sizeof(uint32_t),
                     cudaMemcpyHostToDevice, stream);
-                mFMHARunner->run(fmhaParams);
+                mFmhaDispatcher->run(fmhaParams);
```
Comment (symphonylyh) [inline cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:523]: Just to verify if I understand it correctly @PerkzZheng : the new FMHA dispatcher encapsulates the old FMHARunner & trtllmGenRunner, so it can run both fmha_v2 (<= sm90) and trtllm-gen (>=sm100)

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp [hunk 5]
```diff
@@ -948,10 +959,15 @@ int BertAttentionPlugin::initialize() noexcept
         }
 
         // Load kernels from the pre-compiled cubins.
-        mFMHARunner.reset(new FusedMHARunnerV2(fmhaParams));
-
-        // Fall back to unfused MHA kernels if not supported.
-        mEnableContextFMHA = mFMHARunner->isFmhaSupported();
+        // // The KV input data type. The default is same as dataType.
```
[the remark this answers is not in the document]
Comment (symphonylyh) [reply cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:962]: yes, using fmhadispatcher is correct, but the code comment "Load kernels from the pre-compiled cubins for blackwell" is not? on Hopper it loads hopper cubins, on Blackwell it loads blackwell cubins

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp (hunk 4 above)
Comment (PerkzZheng) [reply cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:523]: yes.

**Review by symphonylyh — APPROVED** [review]

File: cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp [hunk 6]
```diff
@@ -947,11 +958,15 @@ int BertAttentionPlugin::initialize() noexcept
             fmhaParams.saveSoftmax = true;
         }
 
-        // Load kernels from the pre-compiled cubins.
-        mFMHARunner.reset(new FusedMHARunnerV2(fmhaParams));
+        // The KV input data type. The default is same as dataType.
+        fmhaParams.dataTypeKv = data_type;
+        fmhaParams.forceFp32Acc = false;
```
Comment (PerkzZheng) [inline cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:963]: should this be equal to `mFMHAForceFP32Acc` ?

**Review by PerkzZheng — APPROVED** [review]
Others LGTM.
